test/LanguageServer.Engine.Tests/TestProject.cs

```diff
@@ -0,0 +1,386 @@
+            if (testClassType == null)
+                throw new ArgumentNullException(nameof(testClassType));
+
+            if (relativePathSegments == null)
+                throw new ArgumentNullException(nameof(relativePathSegments));
+
+            string projectFileName = GetProjectFile(testClassType, relativePathSegments);
+
+            string projectFileContent = File.ReadAllText(projectFileName);
+            XmlDocumentSyntax projectXml = Parser.ParseText(projectFileContent);
+
+            var xmlPositions = new TextPositions(projectFileContent);
+            var xmlLocator = new XmlLocator(projectXml, xmlPositions);
+
+            return new TestProjectXml(projectXml, xmlLocator, xmlPositions);
+        }
+
+        /// <summary>
+        ///     Get the fully-qualified path of a test project file.
+        /// </summary>
+        /// <typeparam name="TTestClass">
+        ///     The test-suite class used to determine the assembly containing the currently-running test (which can be used to find the test's deployment directory).
+        /// </typeparam>
+        /// <param name="relativePathSegments">
+        ///     The project file's relative (to the test deployment directory) path segments.
+        /// </param>
+        /// <returns>
+        ///     The full path to the project file.
+        /// </returns>
+        public static string GetProjectFile<TTestClass>(params string[] relativePathSegments)
+            where TTestClass : class
+        {
+            if (relativePathSegments == null)
+                throw new ArgumentNullException(nameof(relativePathSegments));
+
+            Type testClassType = typeof(TTestClass);
+
+            return GetProjectFile(testClassType, relativePathSegments);
+        }
+
+        /// <summary>
+        ///     Get the fully-qualified path of a test project file.
+        /// </summary>
+        /// <param name="testClassType">
+        ///     The type of test-suite class used to determine the assembly containing the currently-running test (which can be used to find the test's deployment directory).
+        /// </param>
+        /// <param name="relativePathSegments">
+        ///     The project file's relative (to the test deployment directory) path segments.
+        /// </param>
+        /// <returns>
+        ///     The full path to the project file.
+        /// </returns>
+        public static string GetProjectFile(Type testClassType, params string[] relativePathSegments)
+        {
+            if (testClassType == null)
+                throw new ArgumentNullException(nameof(testClassType));
+
+            if (relativePathSegments == null)
+                throw new ArgumentNullException(nameof(relativePathSegments));
+
+            string testDeploymentDirectory = GetTestDeploymentDirectory(testClassType);
+
+            string projectFileName = Path.Combine(
+                testDeploymentDirectory,
+                Path.Combine(relativePathSegments)
+            );
```

Comment:
I think you can do something like this:
```suggestion
            string projectFileName = Path.Combine([testDeploymentDirectory, .. relativePathSegments]);
```

Reply:
Yeah I originally had that but it felt a little too much like employing a new language feature just for the sake of it when it didn’t make the code that much easier to follow. But I’ll take another look and see if it looks different in the morning 🙂

Reply:
Yeah, I think this was mainly due to my crappy design of the original API (overuse of `params`). Does need fixing, I agree :)